// @wordpress/block-editor/build/components/inspector-controls/block-support-tools-panel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BlockSupportToolsPanel;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _data = require("@wordpress/data");

var _store = require("../../store");

var _utils = require("../../hooks/utils");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function BlockSupportToolsPanel(_ref) {
  let {
    children,
    group,
    label
  } = _ref;
  const {
    clientId,
    attributes
  } = (0, _data.useSelect)(select => {
    const {
      getBlockAttributes,
      getSelectedBlockClientId
    } = select(_store.store);
    const selectedBlockClientId = getSelectedBlockClientId();
    return {
      clientId: selectedBlockClientId,
      attributes: getBlockAttributes(selectedBlockClientId)
    };
  }, []);
  const {
    updateBlockAttributes
  } = (0, _data.useDispatch)(_store.store);

  const resetAll = function () {
    let resetFilters = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
    const {
      style
    } = attributes;
    let newAttributes = {
      style
    };
    resetFilters.forEach(resetFilter => {
      newAttributes = { ...newAttributes,
        ...resetFilter(newAttributes)
      };
    }); // Enforce a cleaned style object.

    newAttributes = { ...newAttributes,
      style: (0, _utils.cleanEmptyObject)(newAttributes.style)
    };
    updateBlockAttributes(clientId, newAttributes);
  };

  return (0, _element.createElement)(_components.__experimentalToolsPanel, {
    className: `${group}-block-support-panel`,
    label: label,
    resetAll: resetAll,
    key: clientId,
    panelId: clientId,
    hasInnerWrapper: true,
    shouldRenderPlaceholderItems: true // Required to maintain fills ordering.

  }, children);
}
