// @wordpress/components/build-module/color-picker/hex-input.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { colord } from 'colord';
/**
 * WordPress dependencies
 */

import { __ } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import { Text } from '../text';
import { Spacer } from '../spacer';
import { space } from '../ui/utils/space';
import { ColorHexInputControl } from './styles';
export const HexInput = ({
  color,
  onChange,
  enableAlpha
}) => {
  const handleValidate = value => {
    if (!colord('#' + value).isValid()) {
      throw new Error('Invalid hex color input');
    }
  };

  return createElement(ColorHexInputControl, {
    prefix: createElement(Spacer, {
      as: Text,
      marginLeft: space(3.5),
      color: "blue",
      lineHeight: 1
    }, "#"),
    value: color.toHex().slice(1).toUpperCase(),
    onChange: nextValue => {
      onChange(colord('#' + nextValue));
    },
    onValidate: handleValidate,
    maxLength: enableAlpha ? 8 : 6,
    label: __('Hex color'),
    hideLabelFromVision: true
  });
};
