// @wordpress/components/build/color-picker/hex-input.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HexInput = void 0;

var _element = require("@wordpress/element");

var _colord = require("colord");

var _i18n = require("@wordpress/i18n");

var _text = require("../text");

var _spacer = require("../spacer");

var _space = require("../ui/utils/space");

var _styles = require("./styles");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const HexInput = ({
  color,
  onChange,
  enableAlpha
}) => {
  const handleValidate = value => {
    if (!(0, _colord.colord)('#' + value).isValid()) {
      throw new Error('Invalid hex color input');
    }
  };

  return (0, _element.createElement)(_styles.ColorHexInputControl, {
    prefix: (0, _element.createElement)(_spacer.Spacer, {
      as: _text.Text,
      marginLeft: (0, _space.space)(3.5),
      color: "blue",
      lineHeight: 1
    }, "#"),
    value: color.toHex().slice(1).toUpperCase(),
    onChange: nextValue => {
      onChange((0, _colord.colord)('#' + nextValue));
    },
    onValidate: handleValidate,
    maxLength: enableAlpha ? 8 : 6,
    label: (0, _i18n.__)('Hex color'),
    hideLabelFromVision: true
  });
};

exports.HexInput = HexInput;
